File: libs/java/common/src/main/java/arda/common/util/MoneyUtils.java
package arda.common.util;

import java.math.BigDecimal;
import java.math.RoundingMode;
import java.text.NumberFormat;
import java.util.Locale;

/**
 * Utility for Money handling, synchronized with Go's decimal handling approach.
 * We prefer using String or Long (minor units) for transport and BigDecimal for calculation.
 */
public class MoneyUtils {
    private static final int DEFAULT_SCALE = 2;
    private static final RoundingMode DEFAULT_ROUNDING = RoundingMode.HALF_UP;

    public static BigDecimal toBigDecimal(String amount) {
        if (amount == null || amount.isBlank()) return BigDecimal.ZERO;
        return new BigDecimal(amount).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING);
    }

    public static String toString(BigDecimal amount) {
        if (amount == null) return "0.00";
        return amount.setScale(DEFAULT_SCALE, DEFAULT_ROUNDING).toPlainString();
    }

    public static String format(BigDecimal amount) {
        return format(amount, new Locale("vi", "VN"));
    }

    public static String format(BigDecimal amount, Locale locale) {
        NumberFormat formatter = NumberFormat.getCurrencyInstance(locale);
        return formatter.format(amount);
    }

    public static BigDecimal add(BigDecimal a, BigDecimal b) {
        return a.add(b).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING);
    }

    public static BigDecimal subtract(BigDecimal a, BigDecimal b) {
        return a.subtract(b).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING);
    }
}

File: arda-be-java/libs/common/src/main/kotlin/arda/common/util/MoneyUtils.kt
package arda.common.util

import java.math.BigDecimal
import java.math.RoundingMode
import java.text.NumberFormat
import java.util.*

/**
 * Utility for Money handling, synchronized with Go's decimal handling approach.
 * We prefer using String or Long (minor units) for transport and BigDecimal for calculation.
 */
object MoneyUtils {
    private const val DEFAULT_SCALE = 2
    private val DEFAULT_ROUNDING = RoundingMode.HALF_UP

    fun toBigDecimal(amount: String?): BigDecimal {
        if (amount.isNullOrBlank()) return BigDecimal.ZERO
        return BigDecimal(amount).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING)
    }

    fun toString(amount: BigDecimal?): String {
        if (amount == null) return "0.00"
        return amount.setScale(DEFAULT_SCALE, DEFAULT_ROUNDING).toPlainString()
    }

    fun format(amount: BigDecimal, locale: Locale = Locale("vi", "VN")): String {
        val formatter = NumberFormat.getCurrencyInstance(locale)
        return formatter.format(amount)
    }

    fun add(a: BigDecimal, b: BigDecimal): BigDecimal = a.add(b).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING)

    fun subtract(a: BigDecimal, b: BigDecimal): BigDecimal = a.subtract(b).setScale(DEFAULT_SCALE, DEFAULT_ROUNDING)
}
